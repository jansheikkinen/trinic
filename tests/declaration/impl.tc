// Impl

struct TestStruct
  x, y: int8
end

trait TestTrait
  function testFunction(self: *Self) -> Self
end

impl TestTrait for TestStruct
  function testFunction(self: *Self) -> Self do
    return(self);
  end
end
