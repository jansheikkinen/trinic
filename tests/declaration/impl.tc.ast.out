
### AST GENERATION ###
Declaration STRUCT
  Arg IDENTIFIER
    Identifier_Arg x
    Identifier_Arg y
      Type TYPE INT8
Declaration TRAIT TestTrait
  Declaration FUNCTION  testFunction
    Arg IDENTIFIER
      Identifier_Arg self
        Type POINTER
          Type IDENTIFIER Self
    Type IDENTIFIER Self
Declaration IMPL
  Type IDENTIFIER TestTrait
  Type IDENTIFIER TestStruct
  Declaration FUNCTION  testFunction
    Arg IDENTIFIER
      Identifier_Arg self
        Type POINTER
          Type IDENTIFIER Self
    Type IDENTIFIER Self
    Stmt BUILTIN return
      Expr VARIABLE self
